### PR TITLE
add jsdoc comments and annotations

### PR DESCRIPTION
**Pull request by wjordan**
This PR adds a bunch of typed annotations to our JSDoc comments, in order to help trace through the sprawling JS objects scattered throughout the method-chains in our frontend apps.

JSDoc comment annotations allow type-sensitive autocomplete-assist in IDEs like the JetBrains suite (RubyMine/WebStorm, etc), without requiring any new language syntax or special [com|trans]pile step (cf. flow/typescript).

For complex, untyped, many-purposed objects like the `report` object used in the milestone-post flow or the `window.appOptions` object used for various quasi-global properties within the app, adequate type annotations are a prerequisite to refactoring, since it's near-impossible to understand the movement of these untyped data objects through our codebase otherwise.

It will be an ongoing task to keep these annotations in sync when the underlying objects change, and probably a near-impossible task to fully flesh out all type annotations throughout the existing codebase, so this PR only offers a partial, incremental effort.

## Example

Before:

image

After:

image

### CODE REVIEW

**Comment by Bjvanminnen** [comment]
This is great - will certainly be useful. It feels somewhat unfortunate to do all this work in documenting these and not getting any validation.

I'm not sure if there are any easy ways to also get validation. A while back I had tried to add validation for some of the reporting stuff (https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/code-studio/reporting.js#L72). It is still quite imperfect, and warns in lots of cases.

In any case, this is a great step in the right direction.

File: apps/src/code-studio/appOptions.js [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+/**
+ * Object representing everything in window.appOptions.
+ *
+ * @typedef {Object} AppOptionsConfig
+ * @property {boolean} embedded
+ * @property {string} scriptName
+ * @property {string} stagePosition
+ * @property {string} levelPosition
+ * @property {AutoplayVideo} autoplayVideo
+ * @property {?} initialAnimationList
```
Comment (islemaster) [inline apps/src/code-studio/appOptions.js]: Type [`SerializedAnimationList`](https://github.com/code-dot-org/code-dot-org/blob/5484df2adb7d2a16a23e4fcdcad250744ce36a54/apps/src/gamelab/PropTypes.js#L132-L136).

File: apps/src/code-studio/appOptions.js [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+ * @property {string} levelGameName
+ * @property {string} skinId
+ * @property {string} baseUrl
+ * @property {string} app
+ * @property {?} droplet
+ * @property {?} pretty
```
Comment (islemaster) [inline apps/src/code-studio/appOptions.js]: I'm having trouble finding this `pretty` property anywhere.

File: apps/src/code-studio/appOptions.js [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+ * @property {Level|Artist|Blockly} level
+ * @property {?} showUnusedBlocks
```
Comment (islemaster) [inline apps/src/code-studio/appOptions.js]: `boolean`

**Comment by wjordan** [comment]
> Not sure how much feedback you want on particular type annotations.

Anything off the top of your (or anyone else's) head would definitely be helpful to get in this doc pass while I'm working on it!

File: apps/src/code-studio/appOptions.js (hunk 2 above)
Comment (wjordan) [reply apps/src/code-studio/appOptions.js]: `appOptions.pretty` is currently set in [`Blockly#blockly_app_options`](https://github.com/code-dot-org/code-dot-org/blob/8474abcc65ae0f8e0d2d8a548120358345f686c4/dashboard/app/models/levels/blockly.rb#L224). It was originally introduced in #1064. Don't know if it's still being referenced anywhere.

**Pull request merged**